Tiles, WIP: improve layout spacing on main dungeon view

aidanholm opened the pull request:
General tiles UI streamlining:
 - The gridded tabs in the sidebar are now flush with the right side of the window.
 - The sidebar now has fixed width, and is always flush against the right side of the window.
 - The main dungeon view expands to fill all remaining space, showing parts of partially visible tiles.
 - The messages box sticks to the bottom of the screen.

There's some other fixes/enhancements as well. Some of them aren't really tiles related and should perhaps be in a separate PR...

rawlins commented on crawl-ref/source/command.cc:
```diff
@@ -796,14 +796,13 @@ static void _add_insert_commands(column_composer &cols, const int column,
 
 static void _add_insert_commands(column_composer &cols, const int column,
-                                 const string desc, const int first, ...)
+                                 const string desc, ...)
```
This was a very good find -- it's bad enough that |amethyst went ahead and implemented a much more drastic solution in 022a3cd6c2ec4970f0af811bda2f5684c8d63321 that eliminates the use of varargs altogether.  (As far as we can tell, nargs is a preemptive defense against the fact that it's very easy to use these versions of _add_insert_commands wrong and not terminate the varargs lists, in fact in a bunch of cases it was doing so.). There's lots more detailed discussion in the devlog from 10/16.

aidanholm replied:
My preferred protection against not terminating varargs in this case would be GCC's `__attribute__ ((sentinel))`, but in C++ NULL != 0, so it caused a bunch of warnings when I tried it.

rawlins replied:
Yeah, I think from the c++ perspective any use of varargs is considered not great, so |amethyst just converted it to a vector.

aidanholm replied:
Ah, found the IRC logs, which for some reason my IRC bouncer isn't working properly... I don't remember this causing a crash, I think it just popped up in valgrind while I was trying to track down one of the other resize issues. But in any case, glad it's fixed :)

rawlins commented on crawl-ref/source/fontwrapper-ft.cc:
```diff
@@ -86,12 +104,11 @@ bool FTFontWrapper::load_font(const char *font_name, unsigned int font_size,
         die_noline("Could not read font '%s'\n", font_name);
     unsigned long size = file_size(f);
-    FT_Byte *ttf = (FT_Byte*)malloc(size);
+    ttf = (FT_Byte*)malloc(size);
```
I realize you're just editing code that was already there, but while you're in there I wonder if this malloc/free pair could be converted to new/delete?  There may be some arcane reason why not that I'm missing since I don't know much about freetype (though FT_Byte is probably just unsigned char), but I think we try to avoid malloc.  Especially given that ttf is now a member variable.

rawlins commented on crawl-ref/source/fontwrapper-ft.cc:
```diff
@@ -56,23 +76,21 @@ FTFontWrapper::~FTFontWrapper()
     delete[] m_glyphs;
     delete[] pixels;
     delete m_buf;
+    if (face) FT_Done_Face(face);
```
tedious comment: should have a linebreak here

rawlins commented on crawl-ref/source/fontwrapper-ft.cc:
```diff
@@ -32,6 +32,24 @@
+    }
+private:
+    FT_Library library;
+    FontLibrary () {
```
tedious comment: `{` should be on a newline given our bracketing conventions (I won't comment all the cases of this but there are a bunch)

rawlins commented on crawl-ref/source/fontwrapper-ft.cc:
```diff
@@ -56,23 +76,21 @@ FTFontWrapper::~FTFontWrapper()
     delete[] pixels;
     delete m_buf;
+    if (face) FT_Done_Face(face);
+    free(ttf);
```
looking at it a bit more I guess this question applies to the already existing m_glyphs and pixels, too

aidanholm replied:
`ttf` is set to null in the constructor, and free is well-defined when called with a null pointer, so it's fine

aidanholm commented:
I've fixed the braces/linebreaks, removed the varargs commit for the issue that's already been fixed in master, and rebased.

I've been playtesting and I've noticed an issue when using a scroll of identify: occasionally the "Identify which item?" screen is blank except for the header text, but I can still select an item with a keypress. Going up/down stairs fixes this and restores correct behavior. I'm not sure if this was caused by my changes, or by other changes in master since 0.21, but I've never noticed it before.

rawlins commented:
I'm getting glitching on OS X, I think because of the clipping stuff; only the lower-left quarter of the dungeon area is shown. This is a retina macbook in case it could be high-DPI-related, here's what it looks like: 

PR: https://www.dropbox.com/s/2dln8dqqoyh8gga/Screenshot%202017-10-17%2016.41.04.png?dl=0
master:

If you don't have access to os x, it might be easier for me to try to debug it myself, but let me know if you have any ideas.

rawlins commented:
Can confirm it's retina-related, things look fine on my mac mini with a normal monitor.

rawlins commented:
Merged, thanks very much for these fixes (and please do more if you want)!  I fixed the retina thing in dc75ea7c3988, it was just a matter of providing glScissor with device pixels rather than logical pixels.
